File: app/src/main/java/com/absinthe/anywhere_/model/ChatQueue.java
package com.absinthe.anywhere_.model;

import androidx.annotation.Nullable;

import com.absinthe.anywhere_.adapter.gift.ChatAdapter;

import java.util.LinkedList;

import timber.log.Timber;

public class ChatQueue extends LinkedList<String> {

    private IChatQueueListener mListener;
    private Thread offerThread;
    private boolean interrupt = false;

    public ChatQueue(IChatQueueListener listener) {
        mListener = listener;
    }

    public boolean offer(String s, int type) {
        boolean result = super.offer(s);
        Timber.d("Chat Enqueue: %s", s);
        mListener.onEnqueue(type);
        return result;
    }

    public void offer(String[] strs) {
        offerThread = new Thread(() -> {
            try {
                for (String str : strs) {
                    if (interrupt) {
                        break;
                    }

                    String prefix = str.substring(0, 3);
                    Timber.d(prefix);
                    switch (prefix) {
                        case GiftChatString.L:
                            offer(str.substring(3), ChatAdapter.TYPE_LEFT);
                            break;
                        case GiftChatString.R:
                            offer(str.substring(3), ChatAdapter.TYPE_RIGHT);
                            break;
                        case GiftChatString.I:
                            offer(str.substring(3), ChatAdapter.TYPE_INFO);
                            break;
                    }
                    Thread.sleep(2000);
                }
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        });
        offerThread.start();
    }

    @Nullable
    @Override
    public String poll() {
        String head = super.poll();
        mListener.onDequeue(head);
        Timber.d("Chat Dequeue: %s", head);
        return head;
    }

    public void stopOffer() {
        interrupt = true;
        offerThread.interrupt();
    }

    public interface IChatQueueListener {
        void onEnqueue(int type);
        void onDequeue(String head);
    }
}

File: app/src/main/java/com/absinthe/anywhere_/model/manager/ChatQueue.kt
package com.absinthe.anywhere_.model.manager

import com.absinthe.anywhere_.adapter.gift.ChatAdapter
import com.absinthe.anywhere_.constants.GiftChatString
import timber.log.Timber
import java.util.*

class ChatQueue(private val mListener: IChatQueueListener) : LinkedList<String>() {

    private var offerThread: Thread? = null
    private var interrupt = false

    fun offer(strings: Array<String>) {
        offerThread = Thread(Runnable {
            try {
                for (str in strings) {
                    if (interrupt) {
                        break
                    }
                    val prefix = str.substring(0, 3)
                    Timber.d(prefix)
                    when (prefix) {
                        GiftChatString.L -> offer(str.substring(3), ChatAdapter.TYPE_LEFT)
                        GiftChatString.R -> offer(str.substring(3), ChatAdapter.TYPE_RIGHT)
                        GiftChatString.I -> offer(str.substring(3), ChatAdapter.TYPE_INFO)
                    }
                    Thread.sleep(2000)
                }
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }
        })
        offerThread!!.start()
    }

    override fun poll(): String? {
        val head = super.poll()!!
        mListener.onDequeue(head)
        Timber.d("Chat Dequeue: %s", head)
        return head
    }

    fun stopOffer() {
        interrupt = true
        offerThread!!.interrupt()
    }

    private fun offer(s: String?, type: Int): Boolean {
        val result = super.offer(s)
        Timber.d("Chat Enqueue: %s", s)
        mListener.onEnqueue(type)
        return result
    }

    interface IChatQueueListener {
        fun onEnqueue(type: Int)
        fun onDequeue(head: String?)
    }

}